token-syncer/src/token_syncer/syncer.clj

```diff
@@ -0,0 +1,162 @@
+;;
+;;       Copyright (c) 2017 Two Sigma Investments, LP.
+;;       All Rights Reserved
+;;
+;;       THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF
+;;       Two Sigma Investments, LP.
+;;
+;;       The copyright notice above does not evidence any
+;;       actual or intended publication of such source code.
+;;
+(ns token-syncer.syncer
+  (:require [clojure.tools.logging :as log]
+            [plumbing.core :as pc]
+            [token-syncer.waiter :as waiter])
+  (:import (org.eclipse.jetty.client HttpClient)))
+
+(defn retrieve-token->url->token-data
+  "Given a lists tokens and cluster urls, retrieve the token description on each cluster."
+  [^HttpClient http-client cluster-urls all-tokens]
+  (pc/map-from-keys
+    (fn [token]
+      (pc/map-from-keys
+        (fn [cluster-url]
+          (waiter/load-token http-client cluster-url token))
+        cluster-urls))
+    all-tokens))
+
+(defn retrieve-token->latest-description
+  "Given token->cluster-url->token-data, retrieves the latest token description for each token
+   among all the clusters."
+  [token->cluster-url->token-data]
+  (pc/map-from-keys
+    (fn [token]
+      (let [latest-data
+            (reduce
+              (fn [accum-data [cluster-url token-data]]
+                (log/info cluster-url "token-data is" token-data ", accum-data" accum-data)
+                (let [accum-last-update-etag (:last-modified-etag accum-data)
+                      accum-last-update-time (get-in accum-data [:description "last-update-time"])
+                      {:strs [last-update-time] :as cluster-token-description} (:description token-data)]
+                  (if (or (nil? accum-last-update-etag)
+                          (nil? accum-last-update-time)
+                          (and last-update-time (< accum-last-update-time last-update-time)))
+                    {:cluster-url cluster-url
+                     :description cluster-token-description
+                     :last-modified-etag (:last-modified-etag token-data "0")}
+                    accum-data)))
+              {}
+              (token->cluster-url->token-data token))]
+        (log/info "Latest data for" token "is" latest-data)
+        latest-data))
+    (keys token->cluster-url->token-data)))
+
+(defn hard-delete-token-on-all-clusters
+  "Hard-deletes a given token on all clusters."
+  [^HttpClient http-client cluster-urls token last-modified-etag]
+  (log/info "Hard-delete" token "on clusters" cluster-urls)
+  (reduce
+    (fn [cluster-sync-result cluster-url]
+      (->> (try
+             (let [response (waiter/hard-delete-token http-client cluster-url token last-modified-etag)]
+               {:message :successfully-hard-deleted-token-on-cluster
+                :response response})
+             (catch Exception ex
+               (log/error ex "Unable to delete" token "on" cluster-url)
+               {:cause (.getMessage ex)
```

Comment:
We are already logging the exception. It simplifies testing if we only use the message. I will rename the keys in the map to `:code` (instead of current `:message`), `:message` (for the exception message) and `:response`.